meshmode/discretization/connection/face.py

```diff
@@ -310,16 +330,52 @@ def make_face_restriction(actx, discr, group_factory, boundary_tag,
             face_vertex_unit_coordinates = \
                     grp_vertex_unit_coordinates[loc_face_vertices]
 
-            # Find A, b such that A [e_1 e_2] + b = [r_1 r_2]
+            # Find face_basis, face_offset such that
+            # face_basis.T @ [e_1 e_2] + face_offset == [r_1 r_2].
+            #
             # (Notation assumes that the volume is 3D and the face is 2D.
             # Code does not.)
 
-            b = face_vertex_unit_coordinates[0]
-            A = (  # noqa
+            face_offset = face_vertex_unit_coordinates[0]
+            face_basis = (
                     face_vertex_unit_coordinates[1:]
-                    - face_vertex_unit_coordinates[0]).T
+                    - face_offset)
+
+            if isinstance(mgrp, SimplexElementGroup):
+                # no further action required
+                pass
+
+            elif isinstance(mgrp, TensorProductElementGroup):
+                if mgrp.dim <= 2:
+                    # faces are simplices, no action required
+                    pass
+
+                elif mgrp.dim == 3:
+                    # Drop the trailing basis vector, so that the convex
+                    # combination of the remaining two is a subset of the face.
+                    reduced_face_basis = face_basis[:-1]
+
+                    # FIXME: This makes the tests pass, but I do not understand
+                    # why this reversal is necessary.
+                    reduced_face_basis = reduced_face_basis[::-1]
```

Comment:
I have no idea why this this is needed, but it makes the tests pass. Any ideas?

Reply:
It seems like there's some mismatch in the order. I also got it to work by removing this and transposing the `bdry_unit_nodes`, i.e. `x[d, i, j] -> x[d, j, i]`. I'll look into it a bit more!

Reply:
Actually, the `test_boundary_interpolation` seems to pass just fine regardless, only the vertex consistency check is failing because it transposes vertices.

Reply:
Hm, after stating the obvious a few times, I think I'm reasonably stumped too :\

Reply:
I'll merge this into inducer/meshmode#70 and continue playing around with it there.